### PR TITLE
feat: server config and engine environment variables

### PR DESCRIPTION
**Pull request by skygeo**
this includes:
- move all the engine configuration (ie runner, swf_domain ...) into a ini file that can be provided at started.
   ```slingshot server [-c slingshot-engine.ini] 
- add a slingshot/confspec.ini in order to validate and set default values for the config
- add a new engine 'environment-variables' config section in order to provide environment variables that will be passed to every steps and plugin containers.
- add minimal readme documentation.

### CODE REVIEW

File: slingshot/engine.py [hunk 1]
```diff
@@ -60,17 +66,38 @@ def start(self, backend='local', swf_domain='slingshot', swf_version='1', auth_c
...
-            # TODO: Handle status
             # TODO: Validate output
             if status == StepStatus.completed:
                 output['name'] = data['name']
+                # Pass generic envirnoment variables to every steps
+                for step in output['do']:
+                    print(step)
+                    step['environment'] = dict(step.get('environment', {}), **self.config.environment_variables)
+                for step in output['rollback']:
+                    step['environment'] = dict(step.get('environment', {}), **self.config.environment_variables)
```
Comment (balintzs) [inline slingshot/engine.py]: Same situation in these as well.

File: slingshot/engine.py [hunk 2]
```diff
@@ -60,17 +66,38 @@ def start(self, backend='local', swf_domain='slingshot', swf_version='1', auth_c
...
+        Running the workflow consists on running every step on the plan.
+
+        Every containers, regardless if it's a plugin or step container, get passed a global list of environment
+        variable that can be when starting the engine.
         """
         workflow = Workflow(events=events)
         for data in config['pipeline']:
-            status, output = await Step(**data, description='').run(
+            # extend the plutin environment variables
+            data['environment'] = dict(data.get('environment', {}), **self.config.environment_variables)
```
Comment (chaudyg) [inline slingshot/engine.py]: I think it's what's happening already. Isn't it?

Comment (balintzs) [inline slingshot/engine.py]: To clarify: the values from `data` should be the ultimate truth. Also, when you do this, if both `data` and `self.config` define the same keyword, you'll get "TypeError: type object got multiple values for keyword argument 'foo'"

File: slingshot/engine.py [hunk 3]
```diff
@@ -60,17 +66,38 @@ def start(self, backend='local', swf_domain='slingshot', swf_version='1', auth_c
...
                 pull_image=self.config.pull_image,
                 auth_config=self.config.auth_config,
             )
-            # TODO: Handle status
             # TODO: Validate output
             if status == StepStatus.completed:
                 output['name'] = data['name']
+                # Pass generic envirnoment variables to every steps
+                for step in output['do']:
+                    print(step)
```
Comment (balintzs) [inline slingshot/engine.py]: Remove this

Comment (skygeo) [inline slingshot/engine.py]: done

File: slingshot/configspec.ini [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+	domain = string(default=None)
+	version = string(default=None)
+
+
+[docker-auth-config]
+# docker registry configuration
+	registry = string(default=None)
+	username = string(default=None)
+	password = string(default=None)
+	email = string(default=None)
```
Comment (skygeo) [inline slingshot/configspec.ini]: I like the idea... but don't know how we could implement this.

We call the 'docker login' command before pulling a container image by providing the default registry. I believe this command supports a single registry.

https://github.com/Skyscanner/slingshot-engine/blob/master/slingshot/core/step.py#L89

How do you see this working in a multi registry context?

File: slingshot/engine.py (hunk 2 above)
Comment (balintzs) [inline slingshot/engine.py]: You are overwriting values in `data` with values in `self.config` whereas it should be the other way around.

Comment (skygeo) [inline slingshot/engine.py]: Well spotted. Fixed now ... and I've update the test as well.

File: slingshot/configspec.ini (hunk 4 above)
Comment (balintzs) [inline slingshot/configspec.ini]: Before you pull an image you check if the auth config is given for its registry. If so, you login to that registry.

Comment (balintzs) [inline slingshot/configspec.ini]: I guess it could be a separate feature all together. So don't worry about it now.

**Comment by balintzs** [comment]
LGTM when tests pass.

**Pull request merged**